### PR TITLE
[new release] ssh-agent and ssh-agent-unix (0.4.0)

### PR DESCRIPTION
**Pull request by reynir**
Ssh-agent protocol parser and serialization implementation

- Project page: <a href="https://github.com/reynir/ocaml-ssh-agent/">https://github.com/reynir/ocaml-ssh-agent/</a>

##### CHANGES:

**breaking changes**
* Sexp has been removed.
* New ssh-ed25519 support, and new constructors `Ssh_agent.Pubkey.Ssh_ed25519` and `Ssh_agent.Privkey.Ssh_ed25519`.

* Support for signing with ed25519 keys.

### CODE REVIEW

File: packages/ssh-agent/ssh-agent.0.4.0/opam [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+dev-repo: "git+https://github.com/reynir/ocaml-ssh-agent.git"
+homepage: "https://github.com/reynir/ocaml-ssh-agent/"
+bug-reports: "https://github.com/reynir/ocaml-ssh-agent/issues/"
+license: "BSD-2-clause"
+build: [
+  ["dune" "build" "-p" name "-j" jobs]
+  ["dune" "runtest" "-p" name "-j" jobs] {with-test}
+]
+depends: [
+  "ocaml" {>= "4.04.0"}
```
Comment (mseri) [inline packages/ssh-agent/ssh-agent.0.4.0/opam:13]:
```suggestion
  "ocaml" {>= "4.08.0"}
```
Due to
```
# (cd _build/default && .ppx/dddc94fd80d07bf5980184f4fc0c28b4/ppx.exe --cookie 'library-name="ssh_agent"' -o src/parse.pp.ml --impl src/parse.ml --dump-ast)
# Fatal error: exception File "src/parse.ml", line 197, characters 2-6: let operators are not supported before OCaml 4.08
#     ocamlopt src/.ssh_agent.objs/native/ssh_agent__Serialize.{cmx,o}
```

Comment (reynir) [reply packages/ssh-agent/ssh-agent.0.4.0/opam]: I think that error is from ppxlib. I think the solution is to add `"ppxlib" {< "version-that-upgraded-to-ast4.08"}` to `conflicts:` (whatever that version is)

File: packages/ssh-agent/ssh-agent.0.4.0/opam [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  "faraday" {>= "0.6"}
+  "mirage-crypto"
+  "mirage-crypto-pk"
+  "mirage-crypto-ec"
+  "cstruct"
+  "alcotest" {with-test}
+]
+conflicts: [
+  "result" {< "1.5"}
+  "ppxlib" {< "0.9.0"}
```
Comment (mseri) [inline packages/ssh-agent/ssh-agent.0.4.0/opam:26]: Feels like this should be a bound in ppxlib itself that is broken?

Comment (mseri) [reply packages/ssh-agent/ssh-agent.0.4.0/opam:26]: No, maybe not, ppxlib is actually compatible, it is the feature that clashes

**Comment by reynir** [comment]
I'm not sure what package is causing this failure right now. ppxlib 0.9.0 should support ocaml 4.08.0 syntax. Maybe I need cstruct>=6.0.0?

**Comment by mseri** [comment]
@kit-ty-kate do you have any idea what could be the cause of
```
#=== ERROR while compiling ssh-agent.0.4.0 ====================================#
# context              2.1.2 | linux/x86_64 | ocaml-base-compiler.4.12.1 | pinned(https://github.com/reynir/ocaml-ssh-agent/releases/download/v0.4.0/ssh-agent-0.4.0.tbz)
# path                 ~/.opam/4.12/.opam-switch/build/ssh-agent.0.4.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ssh-agent -j 31
# exit-code            1
# env-file             ~/.opam/log/ssh-agent-11316-0ffb97.env
# output-file          ~/.opam/log/ssh-agent-11316-0ffb97.out
### output ###
# File "/home/opam/.opam/4.12/lib/bigstringaf/bigstringaf.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# File "/home/opam/.opam/4.12/lib/faraday/faraday.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# File "/home/opam/.opam/4.12/lib/ppx_derivers/ppx_derivers.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
#     ocamlopt src/.ssh_agent.objs/native/ssh_agent__Serialize.{cmx,o}
# File "_none_", line 1:
# Warning 58 [no-cmx-file]: no cmx file was found in path for module Z, and its interface was not compiled with -opaque
#          ppx src/parse.pp.ml (exit 2)
# (cd _build/default && .ppx/dddc94fd80d07bf5980184f4fc0c28b4/ppx.exe --cookie 'library-name="ssh_agent"' -o src/parse.pp.ml --impl src/parse.ml --dump-ast)
# Fatal error: exception File "src/parse.ml", line 197, characters 2-6: let operators are not supported before OCaml 4.08
```
in the lower bounds tests. Seems like the ppxlib bound did not resolve it

**Comment by reynir** [comment]
The culprit was indeed ppx_cstruct. Prior to 6.0.0 they used ppx_tools_versioned so the ppxlib constraint didn't matter there, and before 5.2.0 they used OCaml AST 4.04 and then switched to 4.11.
https://github.com/mirage/ocaml-cstruct/blob/main/CHANGES.md#v520-2020-06-10

**Comment by kit-ty-kate** [comment]
Thanks

**Pull request merged**